### ISSUE TITLE
Some metric doesn't get value (and should not be zero)

### ISSUE THREAD

**Issue by mhow2** [Bitergia/prosoul#186]
Hi @dlumbrer and @valeriocos,

Today I was checking and refining the assessment of our metrics.
In the following example, I'm working on project `Lutece`.
This is the debug dashboard that shows that the `MHFJavaHistoric_filtered_gt10` metricId and we see there are values available in ES index `scava-metrics`.

![2019-10-29_08-46](https://user-images.githubusercontent.com/9987241/67747620-6d8bea80-fa29-11e9-8381-8618bef99e92.png)

Now when I assess in Prosoul, I don't see any raw value for the metric:
![2019-10-29_08-48](https://user-images.githubusercontent.com/9987241/67747723-a4fa9700-fa29-11e9-9fc4-4e0117d62d72.png)

Here is the metric definition for the sake of detail:
![2019-10-29_08-48_1](https://user-images.githubusercontent.com/9987241/67747734-aaf07800-fa29-11e9-83a8-f3aae1d60926.png)

FYI the same issue happens with some others metrics like `bugs.unansweredBugs` 

Now if I look into project `sat4j`, it's working as expected in ProSoul. So I don't understand what's the difference because the `MHFJavaHistoric_filtered_gt10` or `bugs.unansweredBugs` metrics values are available for both projects in ES (and to most of the project of our ES instance btw).

This is the first issue I have.

Now, the second issue I have is the score value that is assessed and taken into account in the screenshot above looks to be `0` : but if we look in ES index `scava-metrics_scores`,  the attribute `Quality_Metrics_CMR` is not present (for project `lutece`) so I tend to understand that from ES perspective the data about these metrics are not recorded at all, so:

From ProSoul perspective, I would like to discuss to possibility not to put score of `0` if the raw metric cannot be found/assessed, to me it looks wrong.

From ES perspective, I would prefer to have all the metrics defined in the model to be recorded regardless if the raw metric data has been found or not, for example with `NaN` value. so consumers of the data in ES can see if there is a missing value and take the opportunity to check where the problem is located (it can be in the metric-to-data mapping, or anywhere else in the system). Of course it should be checked that this new behavior doesn't brake the Kibiter QM dashboard.

WDYT ?

**Comment by valeriocos** [comment]
> FYI the same issue happens with some others metrics like bugs.unansweredBugs
> 
> Now if I look into project sat4j, it's working as expected in ProSoul. So I don't understand what's the difference because the MHFJavaHistoric_filtered_gt10 or bugs.unansweredBugs metrics values are available for both projects in ES (and to most of the project of our ES instance btw).
> 
> This is the first issue I have.

The PR https://github.com/Bitergia/prosoul/pull/189 fixes this problem. In a nutshell, only the first 10 projects were returned. I tested it locally with the fix and I can see Lutece with the correct data.

> Now, the second issue I have is the score value that is assessed and taken into account in the screenshot above looks to be 0 : but if we look in ES index scava-metrics_scores, the attribute Quality_Metrics_CMR is not present (for project lutece) so I tend to understand that from ES perspective the data about these metrics are not recorded at all, so:
> 
> From ProSoul perspective, I would like to discuss to possibility not to put score of 0 if the raw metric cannot be found/assessed, to me it looks wrong.

I have been checking the code and it seems that the data is correctly calculated (meaning that empty data is not replaced with zero values). Thus, the zero values are probably set at visualization time. Tomorrow I'll double check with @dlumbrer and report.

**Comment by mhow2** [comment]
> The PR #189 fixes this problem. In a nutshell, only the first 10 projects were returned. I tested it locally with the fix and I can see Lutece with the correct data.

Thanks ! :+1: 

> Thus, the zero values are probably set at visualization time. 

Yes, instead of NaN. It could be a display issue in Prosoul (and we need NaN value in ES records as well)

TYVM !

**Comment by valeriocos** [comment]
> (and we need NaN value in ES records as well)

I'm not sure this can be easily achieved

**Comment by dlumbrer** [comment]
Hi @mhow2

> 
> Yes, instead of NaN. Indeed, It could be a display issue in Prosoul.
> 

We already worked on this (issue #141) and we decided to put 0 to the assessment that has no value.

 We can change that, what would you like to have instead of 0?, and what do you mean of "it could be display issue in Prosoul"?

**Comment by mhow2** [comment]
Hi @dlumbrer,

I'll try to reformulate.

I don't remember exactly the context of #141 - but now that I have made a step back, what I mean is **we can't put a score of zero to metrics for which no value can be assessed**.

Just for the context and recall what is the current situation, there are two places where we can see the values from the assessment:
1. In the assessment view / dendogram (it include, metrics value/score, attributes average, goals average)
2. In ES on the behalf of the score index (btw notice it only includes metrics value/score, not average values).

> what would you like to have instead of 0?

If there is no value that can be assessed, it could just be something like `no_value`, `NaN`... but definitively not zero. In both place 1) and 2) above.

> what do you mean of "it could be display issue in Prosoul"?

I'm not sure now that you mentioned #141 ...

What I mean is, _when the metric has no value_, it's **displayed** "0" in Prosoul and the attribute and goal averages that are also displayed in the dendogram takes those values into account. (and again, it's not "zero")

Notice that - as I have wrote in the initial description - while it's displayed a score of "zero" in the dendogram, in the meantime the resulting entry in ES is not recorded at all. So in one hand (UI/display) the average is done taken into account "zero" and in the second hand (ES) the value is not recorded. 

I'm insisting : from ES perspective it's a good thing that the value isn't recorded with "zero" into ES, but it's not consistent with what reports the dendogram (which displays "zero", while it's just "no value available").

So to summarize I would suggest  to, if possible : 
- Display NaN/No_Value instead of zero in Prosoul
- In ES, yet if possible and is compatible with what's being done in kibiter (@valeriocos), record *all* metric defined in the model even if we have an assessed value of NaN. So subsequent process that rely on ES data can **explicitly see** that the metric was there but got no value.

**Comment by valeriocos** [comment]
thank you @mhow2 for the clarification

> Display NaN/No_Value instead of zero in Prosoul

OK

> In ES, yet if possible and is compatible with what's being done in kibiter ...

We are checking it

**Comment by mhow2** [comment]
To drive my point home see this : 

![image](https://user-images.githubusercontent.com/9987241/68474294-4ac9b500-0225-11ea-8375-d53cffeccc15.png)

In the above screenshot, I see that two out of the three metrics show "zero" and that the attribute gets a value of 0.67  (2/3 = 0.67). 
If I'd look into ES, I would find only one record (for ...label_joy metric) and would get a totally different result for the related attribute in Kibiter : a value of `2`

**Comment by valeriocos** [comment]
What should be the attribute value in your opinion?

**Comment by mhow2** [comment]
> What should be the attribute value in your opinion?

I would say `2` but with mentioning that it's incomplete. For example in the UI you could mention `2 (!)`

At the metric level, still in the dendogram and when no value can be assessed, instead of `0`  I would mention `no data` or something similar.

WDYT ? Does it sound logical to you guys ? :)

**Comment by valeriocos** [comment]
> To drive my point home see this :

Maybe your point gets deviating from home :) I'm not sure about the right answer. 

Currently, the value 0.67 may be not precise since the other two attributes don't exist. On the other hand, 2 may not be precise either since it would show that the project performs good on that given attribute. Nevertheless, that value is different from a project that scores 2 having the 3 metrics not empty.

I believe that depending on the context 0.67 or 2 may be good results. Probably the best approximation is to let the user decide. Prosoul could have a check button in the assessment that allows to consider (or not) the NaN values.

@dlumbrer please share your thoughts.

**Comment by dlumbrer** [comment]
From a technical point of view, change the assessment in the dendrogram is easy to do, I mean, change the 0 values (when no data) to "No data" is an easy change. And put an alert if the average is incomplete (putting a ( ! ) or changing the color to red, for instance) can be done as well.


Regarding the checkbox/button, we have to study if it is possible, but I would say that a checkbox in the assessment page in order to define if we want the NaN as 0 or just ignore them is possible.

**Comment by valeriocos** [comment]
Thank you @dlumbrer for the feedback!

If we go for the checkbox, we have then to decide what to store in elasticsearch and this may have an impact on the dashboard.

@mhow2 we can implement what you propose (https://github.com/Bitergia/prosoul/issues/186#issuecomment-551658231), and maybe we can leave the checkbox option as future enhancement.

**Comment by mhow2** [comment]
Yes we can have a talk. I just want to mention again that in ES the values that have zero are not recorded so the attribute score deduced would be 2 in the case above, so my first concern it to have something consistent between ES and UI.
My second thought is we don't want to penalize projects because some metric is missing in the system...
Let the user the choice to decide is among my expectation, of course it would be great.

**Comment by valeriocos** [comment]
When are you available for a call?

**Comment by dlumbrer** [comment]
Hi @mhow2,

@valeriocos updated me about the issue, in terms of the dendrogram I'm going to develop the next feature:

- If there is no data in a leaf of the dendrogram there will be the message `no_value` with a red color.
- The average of the parents will not take into account the `no_value` records
- The average that is not complete (with sons with `no_value`) will have a message in red color, like  `( ! )`


Please let me know if you have any comments about this, thanks!

**Comment by mhow2** [comment]
hi @dlumbrer : that sounds good to me, thanks !
@valeriocos : keep me updated with what we both discussed yesterday and what you decide to implement regarding the data in ES when there is no value.

**Comment by dlumbrer** [comment]
> hi @dlumbrer : that sounds good to me, thanks !


This PR (#191) should fix it, please @valeriocos , check it when you can

**Comment by mhow2** [comment]
@dlumbrer I have redeployed this morning it looks OK to me.

**Comment by mhow2** [comment]
hi @valeriocos ,
This morning I have performed more testing on the latest changes about null values in ES.
There might be a little issue (low priority) : I am issuing an assessment between 2018-01-01 and 2018-31-12, and then I query ES with:
```
GET scava-metrics_all_scores/_search?size=50
{
  "query": {
    "term": {
      "project.keyword": "cliflegacy"
    }
  }
}
```
I get 32 items back (which is expected as there are 32 metrics mapped to the model) with one coming from the `scava-metrics_null_scores` (which is expected too). The issue is with the `start_date`/`end_date` of this item which looks like a quarter timespan rather than for the whole timespan

```
      {
        "_index": "scava-metrics_null_scores",
        "_type": "item",
        "_id": "jVGXgm4BCg3cvu4abmTW",
        "_score": 2.9267395,
        "_source": {
          "raw_value": null,
          "end_date": "2019-01-01",
          "type": "all",
          "goal": "MRL_Quality",
          "project": "cliflegacy",
          "calculation_type": "max",
          "score": null,
          "start_date": "2018-10-01",
          "creation_date": "2019-11-19T07:36:30.606324+00:00",
          "attribute": "Quality_SonarQ_Test_Success",
          "score_Sonar test_success_density": null,
          "metric": "Sonar test_success_density"
        }
```

I would have expected to see 2018-01-01 and 2018-31-12 in all items

**Comment by valeriocos** [comment]
I'll have a look at this issue in the next days and report, thanks @mhow2
